lib/gitgitgadget.ts

```diff
@@ -163,50 +164,43 @@ export class GitGitGadget {
         return true;
     }
 
-    public async submit(gitHubUser: string, gitHubUserName: string,
-                        pullRequestURL: string, description: string,
-                        baseLabel: string, baseCommit: string,
-                        headLabel: string, headCommit: string):
+    // better parms would be to pass prinfo object
+    public async submit(pr: IPullRequestInfo, userInfo: IGitHubUser):
         Promise<string | undefined> {
-        if (!this.isUserAllowed(gitHubUser)) {
-            throw new Error(`Permission denied for user ${gitHubUser}`);
-        }
 
-        const [owner, repo, pullRequestNumber] =
-            GitGitGadget.parsePullRequestURL(pullRequestURL);
-        if (owner !== "gitgitgadget" || repo !== "git") {
-            throw new Error(`Unsupported repository: ${pullRequestURL}`);
+        const description = `${pr.title}\n\n${pr.body}`;
+
+        if (pr.baseOwner !== "gitgitgadget" || pr.baseRepo !== "git") {
+            throw new Error(`Unsupported repository: ${pr.pullRequestURL}`);
         }
 
+        // get metadata in work repo
         const metadata =
-            await this.notes.get<IPatchSeriesMetadata>(pullRequestURL);
+            await this.notes.get<IPatchSeriesMetadata>(pr.pullRequestURL);
         const previousTag = metadata && metadata.latestTag ?
             `refs/tags/${metadata.latestTag}` : undefined;
-        await this.updateNotesAndPullRef(pullRequestNumber, previousTag);
+        // update work repo from base
+        await this.updateNotesAndPullRef(pr.number, previousTag);
 
         const series =
-            await PatchSeries.getFromNotes(this.notes, pullRequestURL,
-                                           description, baseLabel, baseCommit,
-                                           headLabel, headCommit,
-                                           gitHubUserName);
+            await PatchSeries.getFromNotes(this.notes, pr.pullRequestURL,
+                                           description, pr.baseLabel,
+                                           pr.baseCommit, pr.headLabel,
+                                           pr.headCommit, userInfo.name);
 
         const send = async (mail: string): Promise<string> => {
             return await parseHeadersAndSendMail(mail, this.smtpOptions);
         };
         const coverMid =
             await series.generateAndSend(console, send,
                                          this.publishTagsAndNotesToRemote,
-                                         pullRequestURL, new Date());
+                                         pr.pullRequestURL, new Date());
         return coverMid;
     }
 
     protected async updateNotesAndPullRef(pullRequestNumber: number,
                                           additionalRef?: string):
         Promise<string> {
-        if (!await isDirectory(this.workDir)) {
-            await git(["init", "--bare", this.workDir]);
-        }
-
```

Comment:
Is this removal intentional?

Reply:
The init is done in the get method at line 44.  I assume that was added at some point without removing the later one.  I can restore the init if I misunderstood.

Reply:
But that `get()` method is static, while `updateNotesAndPullRef()` is not. Therefore, the latter could be called without having had the benefit of the former guaranteeing that the `workDir` exists...

I think we need to reinstate these lines.

Reply:
The static get() method is used to construct a new instance of GitGitGadget.  This is currently the only way to access the protected constructor so the check would have been done?

Would it be better to have the constructor do the git init and call fetchAndReReadOptions()?  This would remove the git init and fetch and readOptions() from the get() method.  It would ensure the git init has been done (and options are current) if there is a new get() alternative added to create a GitGitGadget.  This would be a separate pull request of course.

Reply:
Ah, you're right, of course, the `get()` method is kind of the constructor: we need this to be `async`, and a constructor cannot be `async`.

So your patch is of course fine as-is, as you patiently explained to me often enough for me to understand. Thanks!